Library/Homebrew/resource.rb

```diff
@@ -232,24 +179,28 @@ def sha256(val)
   end
 
   def url(val = nil, **specs)
-    return @url if val.nil?
+    return @url&.to_s if val.nil?
 
     specs = specs.dup
     # Don't allow this to be set.
     specs.delete(:insecure)
 
-    @url = val
-    @using = specs.delete(:using)
-    @download_strategy = DownloadStrategyDetector.detect(url, using)
-    @specs.merge!(specs)
+    specs[:insecure] = true if @insecure
+
+    @url = URL.new(val, specs)
     @downloader = nil
-    @version = detect_version(@version)
+    @download_strategy = @url.download_strategy
   end
 
   def version(val = nil)
-    return @version if val.nil?
+    return super() if val.nil?
 
-    @version = detect_version(val)
+    @version = case val
+    when String  then Version.create(val)
+    when Version then val
+    else
+      raise TypeError, "version '#{val.inspect}' should be a string"
+    end
```

Comment:
Could this be handled by Sorbet instead?

Reply:
Probably but it wouldn't be enforced in homebrew-core since we don't typecheck there yet.

Reply:
Would `brew readall` be sufficient to catch this? If so, might be nice to remove something like this and make `brew readall` always do Sorbet runtime type checking? If it isn't/wouldn't be enough: at most add `TODO` to investigate later.

Increasingly wondering if we've hit the time to make `HOMEBREW_DEVELOPER` also enable type checking at runtime, too. I've had `HOMEBREW_SORBET_RUNTIME` enabled for a while and everything's working fine. CC @dduugg @reitermarkus for people who might have opinions about this.

Reply:
> Would `brew readall` be sufficient to catch this?

If you mean `brew readall` with Sorbet runtime checking then yes it might do - though we should test this with an intentionally bad formula.

> Increasingly wondering if we've hit the time to make `HOMEBREW_DEVELOPER` also enable type checking at runtime, too. I've had `HOMEBREW_SORBET_RUNTIME` enabled for a while and everything's working fine. 

I'm fine with this. Do you notice any measurable overhead out of interest? The no-checking code is a little hacky so in the long term it might be an option to not bother with that.

Reply:
I also have `HOMEBREW_SORBET_RUNTIME` enabled without issue (after some initial hiccups that I PR'd). I can see if Stripe folks have perf stats that they're willing to share. I vaguely recall a 0%-7% impact across services in the early days, depending on factors like whether there was a hot code path or deeply nested types that had to be unpacked. (IIRC Sorbet also skips some complex runtime validations as a result now, so don't be afraid to write complex types based on that anecdata.)

As a starting point, I suggest enabling logging of type errors rather than the default of raising them. See the Sorbet [docs](https://sorbet.org/docs/tconfiguration) and our current [override](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/standalone/sorbet.rb) code.

Reply:
@dduugg Honestly I think logging is going to be more annoying to end-users. If we do this I think we should just enable them for developers/CI and make them errors.

Reply:
Unless there's strong opposition and people want it gone now: I'll leave a comment saying this can go if typechecking is enabled for taps.